### ISSUE TITLE
Precision@K values higher than 1

### ISSUE THREAD

**Issue by davenovelli** [lyst/lightfm#186]
My understanding is that P@K values above 1 should not be possible, but I've managed to find a way. I'm fairly certain there are some constraints I'm missing in my problem setup, and looks like a good opportunity to update the docs and/or include some new error handling in the code (both of which I'm happy to help out with if I can).

Here's the basic setup, and I'll be happy to dig into details as much as necessary...
- I'm attempting to create a user matching recommender for a dating site. There are 6550 users and I'm treating users as items as well as users.
- I'm encoding right swipes as 1's and everything else as 0's (left or unswiped)
- I randomly assign 20% of all swipes to the test data. Not taking most recent, or ensuring that all users are represented in both train and test data
- I'm using the WARP loss function
- I have ~50 user features that I've extracted from their profile, but for testing purposes I've reduced the feature set down to: age, weight, and height. Weight and height have null values. None of these features have been normalized (is that necessary?)

When I fit the model without features (30 epochs, learning rate: .05) I get:
mean AUC: train 0.86, test 0.87   *** this is kinda odd too, test is always equal or higher
mean P@k: train 0.48, test 0.43

This seems to be pretty good, and when I examine the ranked recommendations they are appropriately all different for each user

If I add my user features (which I've tried both by adding as both user_features and item_features, and also just as user_features), I get:
mean AUC: train 0.88, test 0.89
mean P@k: train 11.41, test 2.96

Seems impossible. And also - every user's top 5 recommendations are out of the same set of 6-7 "items". There's no variation.

Does anything stick out as being obviously wrong conceptually?

**Comment by maciejkula** [comment]
Can you check the scores that are returned? There was a [related issue](https://github.com/lyst/lightfm/commit/6ac857f9edf93ced0b6f22ea70ab4a217dfd903c) where if the same score was returned for every item precision/recall scores were nonsensical.

**Comment by maciejkula** [comment]
Also, you must not use 'null' values (I'm guessing you mean `NaN`?).

**Comment by davenovelli** [comment]
Turns out it was the `NaN`s and the not normalized features. AUC and P@k are now both lower after adding user features, but I haven't done a grid search over hyperparameters yet and assume that should help tun things.

Is it worthwhile adding a check for NaN's and or normalized values, and at least writing that out in verbose mode?

Thanks so much for the quick responses!

**Comment by maciejkula** [comment]
No worries. More checks in https://github.com/lyst/lightfm/pull/188

**Comment by dbalabka** [comment]
@maciejkula I have faced with the same problem but w/o using features. I'm using the latest version of LightFM library (1.15.0).

I have two models:
1. Based on users interactions with items and uses items and users features. This model working fine and precision is correct.
2. Second model is trained on interaction matrix where columns are items and rows are "virtual" users that represent all possible variations of users features. I'm planning to use this model for cold start users (similar idea to https://dl.acm.org/doi/10.1145/2948992.2949010). This model in some cases shows precision higher then 1 (e.g. 20-40).

I reviewed the ranks that are generated by `predict_ranks` and some values are about 20-100. According to documentation ranks values should be from 0 till 1.
I suspect that  it might be related that train matrix became less sparse and there might be a situation when "virtual" user might have interaction with all items.

Also, I didn't find any `NaN` values in train and test interaction matrices.
